File: app/src/main/java/com/thanksmister/iot/mqtt/alarmpanel/ui/views/AlarmDisableView.java
/*
 * <!--
 *   ~ Copyright (c) 2017. ThanksMister LLC
 *   ~
 *   ~ Licensed under the Apache License, Version 2.0 (the "License");
 *   ~ you may not use this file except in compliance with the License. 
 *   ~ You may obtain a copy of the License at
 *   ~
 *   ~ http://www.apache.org/licenses/LICENSE-2.0
 *   ~
 *   ~ Unless required by applicable law or agreed to in writing, software distributed 
 *   ~ under the License is distributed on an "AS IS" BASIS, 
 *   ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
 *   ~ See the License for the specific language governing permissions and 
 *   ~ limitations under the License.
 *   -->
 */

package com.thanksmister.iot.mqtt.alarmpanel.ui.views;

import android.content.Context;
import android.os.CountDownTimer;
import android.os.Handler;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.animation.Animation;
import android.view.animation.RotateAnimation;

import com.thanksmister.iot.mqtt.alarmpanel.R;
import com.todddavies.components.progressbar.ProgressWheel;

import butterknife.Bind;
import timber.log.Timber;

public class AlarmDisableView extends BaseAlarmView {

    @Bind(R.id.countDownProgressWheel)
    ProgressWheel countDownProgressWheel;

    private int displaySeconds;
    private CountDownTimer countDownTimer;

    public AlarmDisableView(Context context) {
        super(context);
    }

    public AlarmDisableView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
    }

    public void onCancel(){
        codeComplete = false;
        enteredCode = "";
        showFilledPins(0);
        if(listener != null) {
            listener.onCancel();
        }
    }

    public void startCountDown(int pendingTime) {
        Timber.d("startCountDown: "+ pendingTime*1000);
        final int divideBy = 360/pendingTime;
        countDownTimer = new CountDownTimer(pendingTime*1000, 1000) {
            @Override
            public void onTick(long millisUntilFinished) {
                displaySeconds = (int) (millisUntilFinished / 1000);
                Animation an = new RotateAnimation(0.0f, 90.0f, 250f, 273f);
                an.setFillAfter(true);
                countDownProgressWheel.setText(String.valueOf(displaySeconds));
                countDownProgressWheel.setProgress(displaySeconds * divideBy);
            }

            @Override
            public void onFinish() {
                Timber.d("Timed up...");
                if(listener != null) {
                    listener.onCancel();
                }
            }
        }.start();
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        if(handler != null) {
            handler.removeCallbacks(delayRunnable);
        }
        if(countDownTimer != null) {
            countDownTimer.cancel();
            countDownTimer = null;
        }
    }

    @Override
    public void setCode(int code) {
        this.code = code;
    }

    @Override
    void reset() {
    }

    @Override
    protected void addPinCode(String code) {
        if (codeComplete)
            return;

        enteredCode += code;

        showFilledPins(enteredCode.length());

        if (enteredCode.length() == MAX_CODE_LENGTH) {
            codeComplete = true;
            handler = new Handler();
            handler.postDelayed(delayRunnable, 500);
        }
    }

    private Runnable delayRunnable = new Runnable() {
        @Override
        public void run() {
            handler.removeCallbacks(delayRunnable);
            validateCode(enteredCode);
        }
    };

    @Override
    protected void removePinCode() {
        if (codeComplete) {
            return;
        }

        if (!TextUtils.isEmpty(enteredCode)) {
            enteredCode = enteredCode.substring(0, enteredCode.length() - 1);
        }

        showFilledPins(enteredCode.length());
    }

    private void validateCode(String validateCode) {
        int codeInt = Integer.parseInt(validateCode);
        if(codeInt == code) {
            if(countDownTimer != null) {
                countDownTimer.cancel();
                countDownTimer = null; 
            }
            if(listener != null) {
                listener.onComplete(code);
            }
        } else {
            codeComplete = false;
            enteredCode = "";
            showFilledPins(0);
            if(listener != null) {
                listener.onError();
            }
        }
    }
}

File: app/src/main/java/com/thanksmister/iot/mqtt/alarmpanel/ui/views/AlarmDisableView.kt
/*
 * <!--
 *   ~ Copyright (c) 2017. ThanksMister LLC
 *   ~
 *   ~ Licensed under the Apache License, Version 2.0 (the "License");
 *   ~ you may not use this file except in compliance with the License. 
 *   ~ You may obtain a copy of the License at
 *   ~
 *   ~ http://www.apache.org/licenses/LICENSE-2.0
 *   ~
 *   ~ Unless required by applicable law or agreed to in writing, software distributed 
 *   ~ under the License is distributed on an "AS IS" BASIS, 
 *   ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
 *   ~ See the License for the specific language governing permissions and 
 *   ~ limitations under the License.
 *   -->
 */

package com.thanksmister.iot.mqtt.alarmpanel.ui.views

import android.content.Context
import android.os.CountDownTimer
import android.text.TextUtils
import android.util.AttributeSet
import android.view.View
import android.view.animation.RotateAnimation
import kotlinx.android.synthetic.main.dialog_alarm_disable.view.*

class AlarmDisableView : BaseAlarmView {

    internal var alarmListener: ViewListener? = null
    private var displaySeconds: Int = 0
    private var countDownTimer: CountDownTimer? = null

    private val delayRunnable = object : Runnable {
        override fun run() {
            handler.removeCallbacks(this)
            validateCode(enteredCode)
        }
    }

    interface ViewListener {
        fun onComplete(code: Int)
        fun onError()
        fun onCancel()
    }

    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

    override fun onFinishInflate() {
        super.onFinishInflate()
    }

    fun setListener(listener: ViewListener) {
        this.alarmListener = listener
    }

    fun setUseFingerPrint(value: Boolean) {
        useFingerprint = value
        if(value) {
            disable_fingerprint_layout.visibility = View.VISIBLE
        } else {
            disable_fingerprint_layout.visibility = View.GONE
        }
    }

    override fun onCancel() {
        codeComplete = false
        enteredCode = ""
        showFilledPins(0)
        alarmListener?.onCancel()
    }

    fun startCountDown(pendingTime: Int) {
        if (pendingTime > 0) {
            val divideBy = 360 / pendingTime
            countDownTimer = object : CountDownTimer((pendingTime * 1000).toLong(), 1000) {
                override fun onTick(millisUntilFinished: Long) {
                    displaySeconds = (millisUntilFinished / 1000).toInt()
                    val an = RotateAnimation(0.0f, 90.0f, 250f, 273f)
                    an.fillAfter = true
                    countDownProgressWheel.setText(displaySeconds.toString())
                    countDownProgressWheel.setProgress(displaySeconds * divideBy)
                }

                override fun onFinish() {
                    alarmListener?.onCancel()
                }
            }.start()
        } else {
            if (alarmListener != null) {
                alarmListener?.onCancel()
            }
        }
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        handler.removeCallbacks(delayRunnable)
        countDownTimer?.cancel()
        countDownTimer = null
    }

    override fun reset() {}

    override fun fingerNoMatch() {
        alarmListener?.onError()
    }

    override fun addPinCode(code: String) {
        if (codeComplete)
            return

        enteredCode += code
        showFilledPins(enteredCode.length)
        if (enteredCode.length == BaseAlarmView.Companion.MAX_CODE_LENGTH) {
            codeComplete = true
            handler.postDelayed(delayRunnable, 500)
        }
    }

    override fun removePinCode() {
        if (codeComplete) {
            return
        }

        if (!TextUtils.isEmpty(enteredCode)) {
            enteredCode = enteredCode.substring(0, enteredCode.length - 1)
        }

        showFilledPins(enteredCode.length)
    }

    private fun validateCode(validateCode: String) {
        val codeInt = Integer.parseInt(validateCode)
        if (codeInt == currentCode) {
            countDownTimer?.cancel()
            countDownTimer = null
            alarmListener?.onComplete(currentCode)
        } else {
            codeComplete = false
            enteredCode = ""
            showFilledPins(0)
            alarmListener?.onError()
        }
    }
}